lib/filterx/filterx-eval.c

```diff
@@ -131,6 +133,52 @@ filterx_eval_format_last_error_location_tag(void)
   return filterx_error_format_location_tag(&context->error);
 }
 
+static inline FilterXEvalContext *
+_get_root_context(FilterXEvalContext *context)
+{
+  FilterXEvalContext *root_ctx = context;
+  while (root_ctx->previous_context)
+    root_ctx = root_ctx->previous_context;
```

Comment:
since this could be executed quite frequently, I'd add a root_context pointer to FilterXEvalContext and have that copied over in filterx_eval_begin_context() so it is readily available.

Reply:
or even better if we just store a pointer to the failure_info array in all child contexts, just like the weak_ref array.